apistar/server/app.py

```diff
@@ -0,0 +1,274 @@
+from apistar.server.validation import VALIDATION_COMPONENTS
+from apistar.server.wsgi import WSGI_COMPONENTS, WSGIEnviron, WSGIStartResponse
+
+
+class App():
+    interface = 'wsgi'
+
+    def __init__(self,
+                 routes,
+                 template_dir=None,
+                 static_dir=None,
+                 schema_url='/schema/',
+                 static_url='/static/',
+                 components=None,
+                 event_hooks=None):
+
+        if static_dir is None:
+            static_url = None
+
+        routes = routes + self.include_extra_routes(schema_url, static_url)
+        self.init_document(routes)
+        self.init_router(routes)
+        self.init_templates(template_dir)
+        self.init_staticfiles(static_url, static_dir)
+        self.init_injector(components)
+        self.init_hooks(event_hooks)
+
+    def include_extra_routes(self, schema_url=None, static_url=None):
+        extra_routes = []
+
+        from apistar.server.handlers import serve_schema, serve_static_wsgi
+
+        if schema_url:
+            extra_routes += [
+                Route(schema_url, method='GET', handler=serve_schema, documented=False)
+            ]
+        if static_url:
+            static_url = static_url.rstrip('/') + '/{+filename}'
+            extra_routes += [
+                Route(static_url, method='GET', handler=serve_static_wsgi, documented=False, standalone=True)
+            ]
+        return extra_routes
+
+    def init_document(self, routes):
+        self.document = generate_document(routes)
+
+    def init_router(self, routes):
+        self.router = Router(routes)
+
+    def init_templates(self, template_dir: str=None):
+        if not template_dir:
+            self.templates = None
+        else:
+            template_globals = {'reverse_url': self.reverse_url}
+            self.templates = Templates(template_dir, template_globals)
+
+    def init_staticfiles(self, static_url: str, static_dir: str=None):
+        if not static_dir:
+            self.statics = None
+        else:
+            self.statics = StaticFiles(static_url, static_dir)
+
+    def init_injector(self, components=None):
+        components = components if components else []
+        components = list(WSGI_COMPONENTS + VALIDATION_COMPONENTS) + components
+        initial_components = {
+            'environ': WSGIEnviron,
+            'start_response': WSGIStartResponse,
+            'exc': Exception,
+            'app': App,
+            'path_params': PathParams,
+            'route': Route
+        }
+        self.injector = Injector(components, initial_components)
+
+    def init_hooks(self, event_hooks=None):
+        if event_hooks is None:
+            event_hooks = []
+
+        self.on_request_functions = [
+            hook.on_request for hook in event_hooks
+            if hasattr(hook, 'on_request')
+        ]
+
+        self.on_response_functions = [self.render_response] + [
```

Comment:
I think this list and the `on_error` list should be reversed so that request and response hooks are always at the same depth in the request stack. I went into more details why [here](https://discuss.apistar.org/t/homing-in-on-0-4/426/9?u=bogdanp).